File: ClassEval_T/java/solutuon/SQLQueryBuilder.java
package org.example;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

public class SQLQueryBuilder {

    public static String select(String table, String columns, Map<String, String> where) {
        if (columns == null) {
            columns = "*";
        }
        return select(table, columns.split(",\\s*"), where);
    }

    public static String select(String table, String[] columns, Map<String, String> where) {
        StringBuilder query = new StringBuilder("SELECT ");
        if (columns.length > 0) {
            query.append(String.join(", ", columns));
        } else {
            query.append("*");
        }
        query.append(" FROM ").append(table);

        if (where != null && !where.isEmpty()) {
            query.append(" WHERE ");
            Set<Map.Entry<String, String>> entrySet = where.entrySet();
            boolean first = true;
            for (Map.Entry<String, String> entry : entrySet) {
                if (!first) {
                    query.append(" AND ");
                }
                query.append(entry.getKey()).append("='").append(entry.getValue()).append("'");
                first = false;
            }
        }
        return query.toString();
    }

    public static String insert(String table, Map<String, String> data) {
        StringBuilder query = new StringBuilder("INSERT INTO ").append(table).append(" (");
        StringBuilder values = new StringBuilder(" VALUES (");

        boolean first = true;
        for (Map.Entry<String, String> entry : data.entrySet()) {
            if (!first) {
                query.append(", ");
                values.append(", ");
            }
            query.append(entry.getKey());
            values.append("'").append(entry.getValue()).append("'");
            first = false;
        }

        query.append(")");
        values.append(")");

        return query.append(values).toString();
    }

    public static String delete(String table, Map<String, String> where) {
        StringBuilder query = new StringBuilder("DELETE FROM ").append(table);

        if (where != null && !where.isEmpty()) {
            query.append(" WHERE ");
            Set<Map.Entry<String, String>> entrySet = where.entrySet();
            boolean first = true;
            for (Map.Entry<String, String> entry : entrySet) {
                if (!first) {
                    query.append(" AND ");
                }
                query.append(entry.getKey()).append("='").append(entry.getValue()).append("'");
                first = false;
            }
        }
        return query.toString();
    }

    public static String update(String table, Map<String, String> data, Map<String, String> where) {
        StringBuilder query = new StringBuilder("UPDATE ").append(table).append(" SET ");

        boolean first = true;
        for (Map.Entry<String, String> entry : data.entrySet()) {
            if (!first) {
                query.append(", ");
            }
            query.append(entry.getKey()).append("='").append(entry.getValue()).append("'");
            first = false;
        }

        if (where != null && !where.isEmpty()) {
            query.append(" WHERE ");
            Set<Map.Entry<String, String>> entrySet = where.entrySet();
            boolean firstWhere = true;
            for (Map.Entry<String, String> entry : entrySet) {
                if (!firstWhere) {
                    query.append(" AND ");
                }
                query.append(entry.getKey()).append("='").append(entry.getValue()).append("'");
                firstWhere = false;
            }
        }
        return query.toString();
    }
}

File: ClassEval_T/cpp/solution/SQLQueryBuilder.cpp
#include <string>
#include <vector>
#include <sstream>

class SQLQueryBuilder {
public:
    static std::string select(const std::string& table, const std::vector<std::string>& columns = { "*" }, const std::vector<std::pair<std::string, std::string>>& where = {});
    static std::string insert(const std::string& table, const std::vector<std::pair<std::string, std::string>>& data);
    static std::string delete_(const std::string& table, const std::vector<std::pair<std::string, std::string>>& where = {});
    static std::string update(const std::string& table, const std::vector<std::pair<std::string, std::string>>& data, const std::vector<std::pair<std::string, std::string>>& where = {});
};


std::string SQLQueryBuilder::select(const std::string& table, const std::vector<std::string>& columns, const std::vector<std::pair<std::string, std::string>>& where) {
    std::ostringstream query;
    if (columns.size() == 1 && columns[0] == "*") {
        query << "SELECT *";
    }
    else {
        query << "SELECT ";
        for (size_t i = 0; i < columns.size(); ++i) {
            query << columns[i];
            if (i < columns.size() - 1) {
                query << ", ";
            }
        }
    }
    query << " FROM " << table;
    if (!where.empty()) {
        query << " WHERE ";
        bool first = true;
        for (const auto& [key, value] : where) {
            if (!first) {
                query << " AND ";
            }
            query << key << "='" << value << "'";
            first = false;
        }
    }
    return query.str();
}

std::string SQLQueryBuilder::insert(const std::string& table, const std::vector<std::pair<std::string, std::string>>& data) {
    std::ostringstream query;
    query << "INSERT INTO " << table << " (";
    for (size_t i = 0; i < data.size(); ++i) {
        query << data[i].first;
        if (i < data.size() - 1) {
            query << ", ";
        }
    }
    query << ") VALUES (";
    for (size_t i = 0; i < data.size(); ++i) {
        query << "'" << data[i].second << "'";
        if (i < data.size() - 1) {
            query << ", ";
        }
    }
    query << ")";
    return query.str();
}

std::string SQLQueryBuilder::delete_(const std::string& table, const std::vector<std::pair<std::string, std::string>>& where) {
    std::ostringstream query;
    query << "DELETE FROM " << table;
    if (!where.empty()) {
        query << " WHERE ";
        for (size_t i = 0; i < where.size(); ++i) {
            query << where[i].first << "='" << where[i].second << "'";
            if (i < where.size() - 1) {
                query << " AND ";
            }
        }
    }
    return query.str();
}

std::string SQLQueryBuilder::update(const std::string& table, const std::vector<std::pair<std::string, std::string>>& data, const std::vector<std::pair<std::string, std::string>>& where) {
    std::ostringstream query;
    query << "UPDATE " << table << " SET ";
    for (size_t i = 0; i < data.size(); ++i) {
        query << data[i].first << "='" << data[i].second << "'";
        if (i < data.size() - 1) {
            query << ", ";
        }
    }
    if (!where.empty()) {
        query << " WHERE ";
        for (size_t i = 0; i < where.size(); ++i) {
            query << where[i].first << "='" << where[i].second << "'";
            if (i < where.size() - 1) {
                query << " AND ";
            }
        }
    }
    return query.str();
}
